### PR TITLE
Actually unmount filesystems when exiting repl/server

### LINKED ISSUE #2246: Fix Precog shutdown
Right now, the actor networks are never shut down, meaning that things like `repl/run` will actually hang when you try to exit if you have initialized mimir (typically by cd'ing into the mount).  This is solvable simply by making `Precog` shut down, but this is a little sticky because `Precog` is an object and not a class.  We'll have to figure out some dependent typing stuff (notably relative to `Mimir#Repr`) to get lifecycle stuff.

### LINKED ISSUE #2288: FileSystem close effect isn't being run
This is a continuation of what I referenced in #2287.  It's not clear if this is a `Precog`-specific issue or something more general, though I tend to think it's more general.  Printlns indicate that the `Mimir` `close` action is not being run when I run `repl/run`, `cd mimir` followed by `exit`.  The `exit` will just hang, because the `Precog` actor system never gets shut down, and as I said, printlns indicate that the `Precog` close effect is never sequenced.

I'm guessing that the other connectors just don't exhibit such dramatic issues when they are not shut down properly.  Either that, or I'm derpy and messed up something.

### PR DESCRIPTION
**Pull request by wemrysi**
Fixes #2288.
Fixes #2246.

### CODE REVIEW

File: repl/src/main/scala/quasar/repl/Main.scala [hunk 1]
```diff
@@ -52,42 +52,48 @@ object Main {
...
-        .enableExport(false)
-        .interruptHook(new InterruptHook {
-          def handleInterrupt(console: Console, action: Action) = {
-            console.getShell.out.println("exit")
-            e.unsafePerformSync
-            console.stop
+  private def driver(f: Command => Free[DriverEff, Unit], e: Task[Unit]): Task[Unit] = {
+    def shutdownConsole(c: Console): Task[Unit] =
+      Task.delay(c.getShell.out.println("Exiting...")) *>
+      e.attempt                                        *>
```
Comment (djspiewak) [inline repl/src/main/scala/quasar/repl/Main.scala]: Conceptual niggle…  Shouldn't this be `>>` instead of `*>`?  Both work, but only because `Task` happens to consider hidden evaluation semantics to be part of its substitutability.  It mightn't work with another monad, even a lawful one.

Comment (wemrysi) [inline repl/src/main/scala/quasar/repl/Main.scala]: Hm, wouldn't that imply that `*>` and `>>` would disagree for such a monad? I haven't walked it myself, but I've heard madness lies down that path.

However, maybe `>>` conveys the intent a bit better, in that we want a monad here, not necessarily just an applicative. I'm fine with changing it.

Comment (djspiewak) [inline repl/src/main/scala/quasar/repl/Main.scala]: It's not a big deal, but I do think it conveys the intent better.

`*>` and `>>` may disagree, modulo the definition of substitutability for a given datatype.  `Task` happens to define substitutability to include threading semantics, which is the only sane definition for `Task` since it is a monad for controlling concurrency and threading semantics, but other monads may not have such a definition.

Basically, what it comes down to is the meaning of `=` in the following law (from the `Monad` docs):

```haskell
(>>) = (*>)
```

Gödel is flipping in his grave.

**Comment by djspiewak** [comment]
This is probably ready for merge, since the build passed modulo the weird marklogic setup issues.

**Comment by drostron** [comment]
Have we been merging in that case? I've been waiting for the author to get a clean build in place.

**Comment by djspiewak** [comment]
I guess that's probably something we should decide as a team.  I've been waiting for build #1 and a plurality of non-1 builds.  Couchbase and Marklogic are so flaky, it's really hard to wait on them for PR merging.

**Comment by alissapajer** [comment]
If there is any chance that a build could fail in backend-build for a legit reason, we should definitely wait for the full travis build of all jobs. 

If we're at the stage where we're merging before the build is fully green, then maybe it's time to actually address the flakiness on travis?

**Comment by drostron** [comment]
Is the CB job flakey? We should file an issue around it if so. I've seen it be fairly consistent recently, but you've been running more builds lately.

**Comment by alissapajer** [comment]
@drostron I don't recall seeing any CB bulids fail recently on travis. And we have #2241 open for ML.

And agreed, if we see CB fail, we should open a ticket, because I don't see any open tickets for CB build failing sporadically.

**Comment by drostron** [comment]
_Restarting ML job_

**Comment by djspiewak** [comment]
@alissapajer There are two problems right now with Marklogic on Travis: the initial docker setup (which, afaik, Nic is looking into), and the fact that some of the integration tests use more memory than the build VMs have available, causing the build to hang (and get ultimately killed).  We can't really do anything about that issue, though polyrepoing will improve the situation considerably since it will be surrounded by a considerably less memory-intensive build.  At the very least, it would no longer block quasar merges.

**Comment by alissapajer** [comment]
@djspiewak so if there are some integration tests enabled in marklogic_xml that are not enabled in marklogic_json, and someone made a change to qscript that broke marklogic_xml, then only the xml build would be affected, and the json build would still pass.

**Comment by alissapajer** [comment]
@nicflores @djspiewak do we have an issue created for the markogic initial docker setup problem that we see on travis? We should create one if we don't.

**Comment by djspiewak** [comment]
@alissapajer Are there any such tests though?  It's a good example, though.  For a qscript-changing PR, I would generally wait for the integration tests to fully pass, but other stuff (e.g. things in mimir), there doesn't seem to be a point.

Agreed that we should fix the underlying situation.  This is more a question of how do we staunch the bleeding in the interim.

**Comment by alissapajer** [comment]
@djspiewak I have seen marklogic integration tests enabled in only one of json or xml.

**Comment by alissapajer** [comment]
A fair rule of thumb could be: If marklogic depends on a project you touched, both marklogic builds must pass before merge.

**Comment by nicflores** [comment]
@alissa, I don't think so, at least not currently. Previously #2201 was
opened to address possibly the same error. Should we just re-open that
#2201?

**Comment by alissapajer** [comment]
@nicflores yeah, if we didn't actually fix #2201 then I'd say we should reopen it.

**Comment by alissapajer** [comment]
@nicflores though actually, are we sure it's that same error? I've mostly (entirely?) seen the out of memory error from marklogic, so maybe we should hold off on reopening #2201 until we're sure it's actually the same problem.
